In [2]:
import numpy as np

docs = ['go india',
        'india india',
        'hip hip hurray',
        'jeetega bhai jeetega india jeetega',
        'bharat mata ki jai',
        'kohli kohli',
        'sachin sachin',
        'dhoni dhoni',
        'modi ji ki jai',
        'inquilab zindabad']

In [1]:
from keras.preprocessing.text import Tokenizer
tokenizer= Tokenizer()


In [3]:
tokenizer.fit_on_texts(docs)

In [5]:
sequences = tokenizer.texts_to_sequences(docs)

In [6]:
sequences

[[9, 1],
 [1, 1],
 [3, 3, 10],
 [2, 11, 2, 1, 2],
 [12, 13, 4, 5],
 [6, 6],
 [7, 7],
 [8, 8],
 [14, 15, 4, 5],
 [16, 17]]

In [7]:
from keras.utils import pad_sequences
sequences = pad_sequences(sequences , padding= 'post')
sequences

array([[ 9,  1,  0,  0,  0],
       [ 1,  1,  0,  0,  0],
       [ 3,  3, 10,  0,  0],
       [ 2, 11,  2,  1,  2],
       [12, 13,  4,  5,  0],
       [ 6,  6,  0,  0,  0],
       [ 7,  7,  0,  0,  0],
       [ 8,  8,  0,  0,  0],
       [14, 15,  4,  5,  0],
       [16, 17,  0,  0,  0]])

In [10]:
from keras.models import Sequential 
from keras.layers import Dense , SimpleRNN , Flatten , Embedding

In [ ]:
model = Sequential()
model.add(Embedding(17,output_dim= 2, input_length= 5 )) # 17 is input words
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 5, 2)              34        
                                                                 
Total params: 34
Trainable params: 34
Non-trainable params: 0
_________________________________________________________________


In [12]:
model.compile('adam', 'accuracy')

In [13]:
pred = model.predict(sequences)
print(pred)

1/1 [==============================] - 0s 135ms/step
[[[ 0.00485003  0.00350912]
  [-0.04376886 -0.01687175]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]]

 [[-0.04376886 -0.01687175]
  [-0.04376886 -0.01687175]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]]

 [[-0.00834237 -0.04818437]
  [-0.00834237 -0.04818437]
  [-0.01732125  0.02884557]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]]

 [[-0.00702246  0.03493749]
  [ 0.02413863 -0.0111111 ]
  [-0.00702246  0.03493749]
  [-0.04376886 -0.01687175]
  [-0.00702246  0.03493749]]

 [[ 0.03634336 -0.03730419]
  [-0.04554807 -0.04973521]
  [ 0.03160748  0.04285801]
  [-0.02465549  0.04964615]
  [ 0.00525854  0.00342111]]

 [[-0.03162301  0.00041841]
  [-0.03162301  0.00041841]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]
  [ 0.00525854  0.00342111]]

 [[-0.00240232 -0.00960426]
  [-0.00240232 -0.00960426]
  [ 0.00525854  0.00342111]
  [ 0.00525

#### ***Understanding Embedding . Lest take first array***

---
[ 0.00485003  0.00350912]

  [-0.04376886 -0.01687175]

  [ 0.00525854  0.00342111]

  [ 0.00525854  0.00342111]

  [ 0.00525854  0.00342111]

---
It is an array of first review i.e [ 9,  1,  0,  0,  0]


  * 9 is vectorized into : [ 0.00485003  0.00350912]

  * 1=   [-0.04376886 -0.01687175] and so on .

  **We can see 0 is vectorized into some non zero value . It crate dense data**

---
##### Now using Embedding in Real DataSet

In [14]:
from keras.datasets import imdb
from keras import Sequential
from keras.layers import Dense, SimpleRNN, Embedding, Flatten

In [15]:
(x_train, y_train), (x_test, y_test) = imdb.load_data()

In [16]:
x_train = pad_sequences(x_train, padding='post', maxlen=150)
x_test = pad_sequences(x_test, padding= 'post', maxlen= 150)

In [17]:
x_train.shape

(25000, 150)

In [21]:
model = Sequential()
model.add(Embedding(10000, output_dim=2, input_length=150))
model.add(SimpleRNN(32, activation='tanh',
          input_shape=(150, 1), return_sequences=False))
model.add(Dense(1, activation='sigmoid'))
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, 150, 2)            20000     
                                                                 
 simple_rnn (SimpleRNN)      (None, 32)                1120      
                                                                 
 dense (Dense)               (None, 1)                 33        
                                                                 
Total params: 21,153
Trainable params: 21,153
Non-trainable params: 0
_________________________________________________________________


In [22]:
model.compile(loss='binary_crossentropy',
              optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5,  validation_data=(x_test, y_test))

Epoch 1/5
782/782 [==============================] - 119s 149ms/step - loss: 0.6916 - accuracy: 0.5161 - val_loss: 0.6891 - val_accuracy: 0.5305
Epoch 2/5
782/782 [==============================] - 110s 141ms/step - loss: 0.6684 - accuracy: 0.5771 - val_loss: 0.6856 - val_accuracy: 0.5438
Epoch 3/5
782/782 [==============================] - 113s 145ms/step - loss: 0.6240 - accuracy: 0.6262 - val_loss: 0.6930 - val_accuracy: 0.5458
Epoch 4/5
782/782 [==============================] - 131s 167ms/step - loss: 0.5578 - accuracy: 0.6917 - val_loss: 0.7291 - val_accuracy: 0.5499
Epoch 5/5
782/782 [==============================] - 114s 146ms/step - loss: 0.4681 - accuracy: 0.7792 - val_loss: 0.6016 - val_accuracy: 0.7253
